### PR TITLE
feature: Parse BFF CSVs (CSV pt. 2)

### PR DESCRIPTION
**Pull request by ShrimpCryptid**
Problem
=======
Part 2 of 3 for #211, "prototype direct spreadsheet import."

Most of the line count is from unit tests I've added to validate the CSV parsing!

*Estimated review size: medium, 20-30 minutes*

Solution
========
- Adds support for parsing BioFile Finder CSV formats by mapping BFF keys to CFE-expected keys.
- Adds a LOT of additional unit tests for CSV parsing behavior.

## Type of change
* New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: src/state/image-dataset/csv-dataset/index.ts [hunk 1]
```diff
@@ -52,18 +67,39 @@ const DEFAULT_COLORS = [
     "#BEE952",
 ];
 
-function isNumeric(value: string): boolean {
-    if (typeof value != "string") {
+function isNumeric(value: string | null): boolean {
+    if (value === null) {
+        return true;
```
[the remark this answers is not in the document]
Comment (ShrimpCryptid) [reply src/state/image-dataset/csv-dataset/index.ts:72]: Renamed to `isNullOrNumericString`! CFE allows data to be either string + null arrays OR number + null arrays.

File: src/state/image-dataset/csv-dataset/index.ts [hunk 2]
```diff
@@ -52,18 +67,39 @@ const DEFAULT_COLORS = [
...
+    return false;
+}
+
+function isNullArray(data: FeatureData): data is null[] {
+    for (let i = 0; i < data.length; i++) {
+        if (data[i] !== null) {
+            return false;
+        }
+    }
+    return true;
```
[the remark this answers is not in the document]
Comment (ShrimpCryptid) [reply src/state/image-dataset/csv-dataset/index.ts:102]: Renamed this to `isValidFeatureArray` and added a comment. Thanks for the suggestion!

File: src/state/image-dataset/csv-dataset/index.ts [hunk 3]
```diff
@@ -252,22 +308,100 @@ class CsvRequest implements ImageDataset {
...
+                key: DEFAULT_GROUPBY_NONE,
+                tooltip: "(None)",
+                options,
+            },
+            data: new Array(csvData.length).fill(0),
+        });
+        this.defaultGroupByFeatureKey = DEFAULT_GROUPBY_NONE;
+    }
+
+    private remapBffKeys = (row: Record<string, string>): void => {
```
Comment (rugeli) [inline src/state/image-dataset/csv-dataset/index.ts:361]: nit: separate out a helper function for checking undefined keys to reduce repetitive `if` statements?
```
if (row[A_KEY] === undefined && row[ANOTHER_KEY] !== undefined) {
row[A_KEY] = row[ANOTHER_KEY];
}
```

File: src/state/image-dataset/csv-dataset/index.ts [hunk 4]
```diff
@@ -252,22 +308,100 @@ class CsvRequest implements ImageDataset {
...
+            },
+        };
+        this.featureInfo.set(DEFAULT_GROUPBY_NONE, {
+            type: FeatureType.DISCRETE,
+            def: {
+                discrete: true,
+                displayName: "(None)",
+                description: "(None)",
+                key: DEFAULT_GROUPBY_NONE,
+                tooltip: "(None)",
```
[the remark this answers is not in the document]
Comment (ShrimpCryptid) [reply src/state/image-dataset/csv-dataset/index.ts:353]: Ah, this is a default placeholder feature? Only the key needs to be referenced in other places in the code, while the rest of the name/description/tooltip are named here only.

File: src/state/image-dataset/csv-dataset/index.ts (hunk 3 above)
Comment (ShrimpCryptid) [reply src/state/image-dataset/csv-dataset/index.ts]: Done!

**Review by frasercl — APPROVED** [review]

**Pull request merged**